src/shared/varlink.c

```diff
@@ -0,0 +1,2351 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <sys/poll.h>
+
+#include "alloc-util.h"
+#include "errno-util.h"
+#include "fd-util.h"
+#include "hashmap.h"
+#include "list.h"
+#include "process-util.h"
+#include "set.h"
+#include "socket-util.h"
+#include "string-table.h"
+#include "string-util.h"
+#include "strv.h"
+#include "time-util.h"
+#include "umask-util.h"
+#include "user-util.h"
+#include "varlink.h"
+
+#define VARLINK_DEFAULT_CONNECTIONS_MAX 4096U
+#define VARLINK_DEFAULT_CONNECTIONS_PER_UID_MAX 1024U
+
+#define VARLINK_DEFAULT_TIMEOUT_USEC (45U*USEC_PER_SEC)
+#define VARLINK_BUFFER_MAX (16U*1024U*1024U)
+#define VARLINK_READ_SIZE (64U*1024U)
+
+typedef enum VarlinkState {
+        /* Client side states */
+        VARLINK_IDLE_CLIENT,
+        VARLINK_AWAITING_REPLY,
+        VARLINK_CALLING,
+        VARLINK_CALLED,
+        VARLINK_PROCESSING_REPLY,
+
+        /* Server side states */
+        VARLINK_IDLE_SERVER,
+        VARLINK_PROCESSING_METHOD,
+        VARLINK_PROCESSING_METHOD_MORE,
+        VARLINK_PROCESSING_METHOD_ONEWAY,
+        VARLINK_PROCESSED_METHOD,
+        VARLINK_PROCESSED_METHOD_MORE,
+        VARLINK_PENDING_METHOD,
+        VARLINK_PENDING_METHOD_MORE,
+
+        /* Common states (only during shutdown) */
+        VARLINK_PENDING_DISCONNECT,
+        VARLINK_PENDING_TIMEOUT,
+        VARLINK_PROCESSING_DISCONNECT,
+        VARLINK_PROCESSING_TIMEOUT,
+        VARLINK_PROCESSING_FAILURE,
+        VARLINK_DISCONNECTED,
+
+        _VARLINK_STATE_MAX,
+        _VARLINK_STATE_INVALID = -1
+} VarlinkState;
+
+#define VARLINK_STATE_IS_CONNECTED(state)               \
+        IN_SET(state,                                   \
+               VARLINK_IDLE_CLIENT,                     \
+               VARLINK_AWAITING_REPLY,                  \
+               VARLINK_CALLING,                         \
+               VARLINK_CALLED,                          \
+               VARLINK_PROCESSING_REPLY,                \
+               VARLINK_PROCESSING_FAILURE,              \
+               VARLINK_IDLE_SERVER,                     \
+               VARLINK_PROCESSING_METHOD,               \
+               VARLINK_PROCESSING_METHOD_MORE,          \
+               VARLINK_PROCESSED_METHOD,                \
+               VARLINK_PROCESSED_METHOD_MORE)
+
+struct Varlink {
+        unsigned n_ref;
+
+        VarlinkServer *server;
+
+        VarlinkState state;
+        bool connecting;
```

Comment:
Do we need both `VARLINK_STATE_IS_CONNECTED()` and `connecting`?

Reply:
I decided to keep these things separate, since otherwise we'd have to duplicate most of the states into a "WHILE_CONNECTING" and a "WHILE_ESTABLISHED" state. The idea behind that is this:
when connecting a socket to a peer asynchronously (i.e. when connect() on an O_NONBLOCk socket returns EINPROGRESS) we are in a special state: everything is great, we should accept messages from the user to be queued already, however we cannot write anything to (or read from) the socket  yet. If we did we'd get ENOTCONN from it. We need to know whether we are in this state so that we can avoid calling read() and write() in that case, because if we did it anway, and then get ENOTCONN, which tells us very little: it could mean the connection is not established yet, or it could mean the connection is not established anymore. This boolean discerns the two states.

I will rename VARLINK_STATE_IS_CONNECTED() to VARLINK_STATE_IS_ALIVE() now, as that hopefuly is a bit more explanatory: after all during VARLINK_STATE_IS_CONNECTED() unlike the name suggested we didn't have to be fully connected yet. "alive" otoh hopefully clears up that things are looking good even though we are not yet connected, but might be later.

Or to say this differently: the VarlinkState is a high-level state that encodes our own idea of the varlink connection, i.e. whether we are a client or a server, and whether we are currently processing, sending or dispatching a method call. The 'connecting' boolean otoh tells us whether the low-level transport is actually ready to operate just yet, or we still need to wait a bit. 

The 'connecting' boolean is set whenever we see EINPROGRESS on connect(), and turned off again whenever we see the first EPOLLOUT, as that's indication that the connection setup is ready. Between these two steps we are happy to accept jobs from the user, but we can't just yet write them to the socket.

I'll add longer comments about this to the sources.